### ISSUE TITLE
Add semantic highlighting support for Dockerfiles

### ISSUE THREAD

**Issue by rcjsuen** [microsoft/vscode-docker#1840]
With the finalization of the [semantic highlighting](https://github.com/microsoft/vscode/wiki/Semantic-Highlighting-Overview) API in [Visual Studio Code 1.44](https://code.visualstudio.com/updates/v1_44#_extension-authoring), the Docker extension can now look to resolving long standing bug reports about malformed syntax highlighting in Dockerfiles using code instead of a TextMate grammar. I have tried to do a sweep of the GitHub issues that have been opened in the past to capture the various requests and bug reports about syntax highlighting in Dockerfiles.

- https://github.com/microsoft/vscode-docker/issues/43 (request to add syntax highlighting for `HEALTHCHECK` flags)
- https://github.com/microsoft/vscode-docker/issues/90 (request to add syntax highlighting for `ARG` and `ENV` variables)
- https://github.com/microsoft/vscode-docker/issues/75#issue-216515077 (bug example)
- https://github.com/microsoft/vscode-docker/issues/75#issuecomment-291277804 (bug example)
- https://github.com/microsoft/vscode-docker/issues/75#issuecomment-338812822 (bug example)
- https://github.com/microsoft/vscode-docker/issues/75#issuecomment-401341827 (bug example)
- https://github.com/microsoft/vscode-docker/issues/245#issuecomment-379113269 (bug example)
- https://github.com/microsoft/vscode-docker/issues/292#issue-337954539 (bug example)

**Open question:**
The extension currently depends on Visual Studio Code 1.41 which was released in December 2019.

https://github.com/microsoft/vscode-docker/blob/95fb88977958ea2eb59f81369c399986e094bc2e/package.json#L2439-L2441

Adding semantic highlighting support will change this dependency to 1.44. Not sure to me how many users would be affected by this. Does Microsoft have telemetry information about this to know what percentage of this extension's users are on which release of Visual Studio Code?

**Comment by bwateratmsft** [comment]
I took a look at the last 24 hours, here's the numbers we have for Docker extension users:

| Version | Machine Count |
| --- | --- |
| 1.44.0 | 31205 |
| 1.43.2 | 76885 |
| 1.43.1 | 3809 |
| 1.43.0 | 3359 |
| 1.42.1 | 3757 |
| 1.41.1 | 2607 |

These count the highest version a machine reported in the last 24 hours, i.e. people who have upgraded will only show up in 1.44.0.

The fact that the overwhelming majority of the numbers are in 1.43.2 or 1.44.0 tells me that nearly everyone stays up to date. I'll keep an eye on these numbers throughout the week, it hasn't even been 24 hours since 1.44.0 released so it can be expected to get even higher.

**Comment by rcjsuen** [comment]
@bwateratmsft The `vscode.Terminal` interface [changed in 1.42](https://github.com/microsoft/vscode/commit/9ebb7c43bc99fd6e1a295040674d1f8e5831b9be) so the `test/TestTerminalProvider.ts` must be updated.

https://github.com/microsoft/vscode-docker/blob/95fb88977958ea2eb59f81369c399986e094bc2e/test/TestTerminalProvider.ts#L36

**Comment by bwateratmsft** [comment]
Incidentally `TerminalProvider` is on its way out the door, #251

**Comment by dbreshears** [comment]
Let's monitor telemetry and wait for adoption of 1.44 to get closer to 70%

**Comment by bwateratmsft** [comment]
We're nearly there already. This is among our extension version 1.0.0 users, last 24 hours:

| Version | Count | % of Total |
| --- | --- | --- |
| 1.44.1 | 33316 | 25.6 |
| 1.44.0 | 64414 | 49.5 |
| 1.43.2 | 20371 | 15.7 |
| 1.43.1 | 2752 | 2.1 |
| 1.43.0 | 2695 | 2.1 |
| 1.42.1 | 3221 | 2.5 |
| 1.42.0 | 687 | .5 |
| 1.41.1 | 2357 | 1.8 |
| 1.41.0 | 310 | .2 |

1.44 already accounts for ~75%.

**Comment by bwateratmsft** [comment]
@rcjsuen If you're ready, go ahead and prep a PR so we can get the change into the release as early as possible in the cycle. Thanks again!

**Comment by rcjsuen** [comment]
@bwateratmsft Sorry, perhaps I should have been more clear when I [commented earlier](https://github.com/microsoft/vscode-docker/issues/1840#issuecomment-611578179).

Can someone fix these compiler errors first (which can be observed simply by changing the dependency to 1.42)?

<img width="805" alt="image" src="https://user-images.githubusercontent.com/15629116/79601333-a5115080-80b6-11ea-9226-0546be277864.png">

**Comment by bwateratmsft** [comment]
Ah, yes, I'll be getting that fixed when I merge https://github.com/microsoft/vscode-docker/tree/bmw/terminals which should be soon, probably early-mid next week.

**Comment by bwateratmsft** [comment]
@rcjsuen You should now be unblocked. I upgraded the engine to 1.44 and fixed the bug (#1891); (though not by removing `ITerminalProvider` as originally planned, which ended up being nightmarishly complicated due to shell quoting issues).

**Comment by rcjsuen** [comment]
@bwateratmsft Thanks! I'll submit a pull request later today or tomorrow to enable semantic highlighting.

**Comment by rcjsuen** [comment]
@bwateratmsft I have opened #1899.

**Comment by bwateratmsft** [comment]
@rcjsuen We can close this now that #1899 is submitted, right?

**Comment by rcjsuen** [comment]
> @rcjsuen We can close this now that #1899 is submitted, right?

@bwateratmsft I think we are good for the most part. Should I open a separate issue regarding the configuration of custom colours or should we just forget about it?

To everyone else, I'll reiterate what was posted in #1899 as a reference for toggling this with the JSON settings editor.

1. How do I enable/disable semantic highlighting?

   ```JSON
   "editor.semanticHighlighting.enabled": true
   ```

1. How do I enable/disable semantic highlighting only for Dockerfiles?

   ```JSON
   "[dockerfile]": {
       "editor.semanticHighlighting.enabled": true
   }
   ```

**Comment by bwateratmsft** [comment]
> Should I open a separate issue regarding the configuration of custom colours or should we just forget about it?

@rcjsuen Yeah, I think that can be separate from this.

**Comment by rcjsuen** [comment]
> > Should I open a separate issue regarding the configuration of custom colours or should we just forget about it?
> 
> @rcjsuen Yeah, I think that can be separate from this.

@bwateratmsft OK. Let me do some research and then open an issue when I have something to share. I think we can close this then.

Thanks for your support and for answering the various questions I've been asking of you all!

**Comment by bwateratmsft** [comment]
Thank you for the language server!

**Comment by bwateratmsft** [comment]
This feature is now available in version 1.2.0 of the [Docker extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker).
